src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionServiceV1.java

```diff
@@ -402,6 +438,147 @@ public void appendMetaColIfNotExist(final String tableName) {
+  }
+
+  /**
+   * Alter table to add columns according to a map from columnNames to their types
+   *
+   * @param tableName the name of the table
+   * @param columnToType the mapping from the columnNames to their types
+   */
+  @Override
+  public void appendColumns(String tableName, Map<String, String> columnToType) {
+    checkConnection();
+    InternalUtils.assertNotEmpty("tableName", tableName);
+    StringBuilder appendColumnQuery = new StringBuilder("alter table identifier(?) add column ");
+    boolean first = true;
+    StringBuilder logColumn = new StringBuilder("[");
+    for (String columnName : columnToType.keySet()) {
+      if (first) {
+        first = false;
+      } else {
+        appendColumnQuery.append(", add column ");
+        logColumn.append(",");
+      }
+      appendColumnQuery
+          .append(columnName)
+          .append(" ")
+          .append(columnToType.get(columnName))
+          .append(" comment 'column created by schema evolution from Snowflake Kafka connector'");
+      logColumn.append(columnName).append(" (").append(columnToType.get(columnName)).append(")");
+    }
+    try {
+      PreparedStatement stmt = conn.prepareStatement(appendColumnQuery.toString());
+      stmt.setString(1, tableName);
+      stmt.execute();
+      stmt.close();
+    } catch (SQLException e) {
+      throw SnowflakeErrors.ERROR_2015.getException(e);
+    }
+
+    logColumn.insert(0, "Following columns created for table {}:\n").append("]");
+    logInfo(logColumn.toString(), tableName);
+  }
+
+  /**
+   * Alter table to drop non-nullability of a list of columns
+   *
+   * @param tableName the name of the table
+   * @param columnNames the list of columnNames
+   */
+  @Override
+  public void alterNonNullableColumns(String tableName, List<String> columnNames) {
+    checkConnection();
+    InternalUtils.assertNotEmpty("tableName", tableName);
+    StringBuilder query = new StringBuilder("alter table identifier(?) alter ");
+    boolean first = true;
+    StringBuilder logColumn = new StringBuilder("[");
+    for (String columnName : columnNames) {
+      if (first) {
+        first = false;
+      } else {
+        query.append(", ");
+        logColumn.append(", ");
+      }
+      query
+          .append(columnName)
+          .append(" drop not null, ")
+          .append(columnName)
+          .append(
+              " comment 'column altered to be nullable by schema evolution from Snowflake Kafka"
+                  + " connector'");
+      logColumn.append(columnName);
+    }
+    try {
+      PreparedStatement stmt = conn.prepareStatement(query.toString());
+      stmt.setString(1, tableName);
+      stmt.execute();
+      stmt.close();
+    } catch (SQLException e) {
+      throw SnowflakeErrors.ERROR_2016.getException(e);
+    }
+
+    logColumn
+        .insert(0, "Following columns' non-nullabilty was dropped for table {}:\n")
+        .append("]");
```

Comment:
?

Reply:
I don't know what is wrong here...

Reply:
was confused with insert(0, "") API. nvm 
can you comment we are adding at beginning